File: HappyDog.Infrastructure.Test/PaginationTests/GetFlexibleLinksTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace HappyDog.Infrastructure.Test.PaginationTests
{
    [TestClass]
    public class GetFlexibleLinksTest
    {
        [TestMethod]
        public void Test0()
        {
            var pagination = new Pagination<int>
            {
                Page = 1,
                Size = 10,
                TotalItems = 10,
                AutoHide = false,
                PreviousText = "&laquo;",
                NextText = "&raquo;"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item active\"><span class=\"page-link\">1</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test1()
        {
            var pagination = new Pagination<int>
            {
                Page = 1,
                Size = 10,
                TotalItems = 10,
                AutoHide = true,
                PreviousText = "&laquo;",
                NextText = "&raquo;"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            Assert.AreEqual(string.Empty, html);
        }

        [TestMethod]
        public void Test2()
        {
            var pagination = new Pagination<int>
            {
                Page = 2,
                Size = 10,
                TotalItems = 11,
                PreviousText = "Previous",
                NextText = "&raquo;"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item active\"><span class=\"page-link\">2</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test3()
        {
            var pagination = new Pagination<int>
            {
                Page = 1,
                Size = 10,
                TotalItems = 48,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "1 / 5"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item active\"><span class=\"page-link\">1</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">1 / 5</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test4()
        {
            var pagination = new Pagination<int>
            {
                Page = 2,
                Size = 10,
                TotalItems = 48,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "2 / 5"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item active\"><span class=\"page-link\">2</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">2 / 5</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test5()
        {
            var pagination = new Pagination<int>
            {
                Page = 3,
                Size = 10,
                TotalItems = 48,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "3 / 5"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item active\"><span class=\"page-link\">3</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">3 / 5</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test6()
        {
            var pagination = new Pagination<int>
            {
                Page = 4,
                Size = 10,
                TotalItems = 48,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "4 / 5"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item active\"><span class=\"page-link\">4</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">4 / 5</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test7()
        {
            var pagination = new Pagination<int>
            {
                Page = 9,
                Size = 10,
                TotalItems = 480,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "9 / 48"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 5);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=8\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=7\">7</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=8\">8</a></li><li class=\"page-item active\"><span class=\"page-link\">9</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=10\">10</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=11\">11</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=10\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">9 / 48</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }

        [TestMethod]
        public void Test8()
        {
            var pagination = new Pagination<int>
            {
                Page = 9,
                Size = 10,
                TotalItems = 48,
                PreviousText = "Previous",
                NextText = "Next",
                Append = "9 / 5"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 3);
            Assert.AreEqual(string.Empty, html);
        }

        [TestMethod]
        public void Test9()
        {
            var pagination = new Pagination<int>
            {
                Page = 4,
                Size = 20,
                TotalItems = 63,
                PreviousText = "Previous",
                NextText = "Next"
            };
            string html = pagination.GetFlexibleLinks(p => "/test?p=" + p, 7);
            string expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item active\"><span class=\"page-link\">4</span></li></ul></nav>";
            Assert.AreEqual(expected, html);
        }
    }
}

File: HappyDog.Java/src/test/java/wang/doghappy/java/util/pagination/GetFlexibleLinksTest.java
package wang.doghappy.java.util.pagination;

import org.junit.jupiter.api.Test;
import wang.doghappy.java.util.Pagination;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class GetFlexibleLinksTest {
    @Test
    public void test0() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(1);
        pagination.setSize(10);
        pagination.setTotalItems(10);
        pagination.setAutoHide(false);
        pagination.setPreviousText("&laquo;");
        pagination.setNextText("&raquo;");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item active\"><span class=\"page-link\">1</span></li><li class=\"page-item\"><span class=\"page-link\">1 / 1</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test1() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(1);
        pagination.setSize(10);
        pagination.setTotalItems(10);
        pagination.setAutoHide(true);
        pagination.setPreviousText("&laquo;");
        pagination.setNextText("&raquo;");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        assertEquals("", html);
    }

    @Test
    public void test2() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(2);
        pagination.setSize(10);
        pagination.setTotalItems(11);
        pagination.setPreviousText("Previous");
        pagination.setNextText("&raquo;");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item active\"><span class=\"page-link\">2</span></li><li class=\"page-item\"><span class=\"page-link\">2 / 2</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test3() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(1);
        pagination.setSize(10);
        pagination.setTotalItems(48);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item active\"><span class=\"page-link\">1</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">1 / 5</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test4() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(2);
        pagination.setSize(10);
        pagination.setTotalItems(48);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item active\"><span class=\"page-link\">2</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">2 / 5</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test5() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(3);
        pagination.setSize(10);
        pagination.setTotalItems(48);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item active\"><span class=\"page-link\">3</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">4</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=4\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">3 / 5</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test6() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(4);
        pagination.setSize(10);
        pagination.setTotalItems(48);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item active\"><span class=\"page-link\">4</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">5</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=5\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">4 / 5</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test7() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(9);
        pagination.setSize(10);
        pagination.setTotalItems(480);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 5);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=8\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=7\">7</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=8\">8</a></li><li class=\"page-item active\"><span class=\"page-link\">9</span></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=10\">10</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=11\">11</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=10\">Next</a></li><li class=\"page-item\"><span class=\"page-link\">9 / 48</span></li></ul></nav>";
        assertEquals(expected, html);
    }

    @Test
    public void test8() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(9);
        pagination.setSize(10);
        pagination.setTotalItems(48);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 3);
        assertEquals("", html);
    }

    @Test
    public void test9() {
        var pagination = new Pagination<Integer>();
        pagination.setPage(4);
        pagination.setSize(20);
        pagination.setTotalItems(63);
        pagination.setPreviousText("Previous");
        pagination.setNextText("Next");
        String html = pagination.getFlexibleLinks(p -> "/test?p=" + p, 7);
        String expected = "<nav><ul class=\"pagination\"><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">Previous</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=1\">1</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=2\">2</a></li><li class=\"page-item\"><a class=\"page-link\" href=\"/test?p=3\">3</a></li><li class=\"page-item active\"><span class=\"page-link\">4</span></li><li class=\"page-item\"><span class=\"page-link\">4 / 4</span></li></ul></nav>";
        assertEquals(expected, html);
    }
}
